### ISSUE TITLE
Ask for pipe support 

### ISSUE THREAD

**Issue by Huangzhiqiang** [ksvc/ks265codec#11]
As title

**Comment by zhangxianguo** [comment]
can you give more details?

**Comment by tnti** [comment]
一个demo连pipeline都要阉割...

**Comment by tnti** [comment]
ffmpeg.exe -i "%~dpn1.mp4" -f rawvideo -an -| AppEncoder_x64.exe -i - -preset veryfast -wdt 1920 -hgt 1080 -fr 24000/1001 -rc 1 -br 2000 -iper 128  -b "%~dpn1.hvc" 

例如 然后ffmepg报pipeline失败啥的  还有 BR那个是bps... 为毛不是kbps 强迫症表示受不了...

**Comment by zhangxianguo** [comment]
@tnti 我们的br参数的单位是kbps的，不过readme中确实没有介绍单位。

**Comment by tnti** [comment]
不过同样X265和你们家的编码器我都用2000kbps差距有点大啊
你们家的是：
Bit rate                       : 253 kb/s
Maximum bit rate               : 378 kb/s
X265的是：
Bit rate                       : 2 005 kb/s
Maximum bit rate               : 2 861 kb/s

pipeline加上去把 把pipeline支持y4m和yuv弄进去省事点 要不每次都要弄一大坨yuv太麻烦
 其他参数定死挺好 demo没必要把所有功能全放出来

**Comment by zhangxianguo** [comment]
能分享一下你用的参数么？感觉是码控没有生效。

**Comment by tnti** [comment]
AppEncoder_x64.exe -i "%~dpn1.yuv" -preset medium -wdt 1920 -hgt 1080 -fr 24000/1001 -rc 1 -br 2000 -iper 128  -b "%~dpn1.hvc" 

avs4x26x.exe --bitrate 2000 --no-open-gop --bframes 6 --rd-refine --weightb --rc-lookahead 50 --ref 5 --weightp --no-rect --no-info --no-ssim --no-psnr --output "%~dpn1.hevc" "%~dpn1.avs"

300帧的玩意 随便切的...  拿HM解的时候看见QP在跳了 应该生效了...
链接：http://pan.baidu.com/s/1kVwINrD 密码：b0q4 附上随手截的YUV...

**Comment by zhangxianguo** [comment]
@tnti 首先说谢谢。我们的appencoder对 -fr命令后的除号不识别。实际上我们限定-fr后面得是小数，这也是我们将来需要改进的。不过24000/1001这种支持实际上是有一定必要性的。我们会考虑在将来的版本添加支持，您在当前测试时可以考虑用23.97

**Comment by zhangxianguo** [comment]
@tnti 关于管道的问题，我们的编码器跟x264和x265的定位毕竟有些不同，我们不是开放库的软件。我会跟团队的人讨论一下，到底是否添加支持。

**Comment by tnti** [comment]
还真是-fr的问题... 用小数码率控制就没事了...

**Comment by Selur** [comment]
I also would agree that pipe support would be nice, but instead fo a rawvideo video a yuv4mpegpipe might be better suited (since y4m comes with headers that already include the resolution). 

An example call could be:
> ffmpeg -loglevel fatal -an -sn -i "F:\test.avi" -r 25  -pix_fmt yuv420p -f yuv4mpegpipe - | AppEncoder_x64.exe -i - -preset veryfast -latency offline -fr 25 -rc 1 -br 20000 -iper 128 -b h:\Output\test.265

 ( -wdt X and -hgt Y could be automatically set based on the y4m header information) 

Having pipe support would be great to avoid having to create humongous temporal yuv files.

**Comment by zhangxianguo** [comment]
@Selur Indeed, pipe support will make it easy to produce 265 streams. However, this support needs Kingsoft 's agreement. Still being discussed. Thanks a lot for your concern.

**Comment by Selur** [comment]
Any update on this?

**Comment by zhangxianguo** [comment]
Pipe support is still not permitted. Any problem if we update an ffmpeg with ksc265 and some usage limitation?

**Comment by Selur** [comment]
No, from my point of view an ffmpeg version would be fine.
Usage limitations also would be fine, at least this way testing the encoder would be possible in a more convenient way. (testing real content and always having to create humongous temporary raw files is a pain ;))

**Comment by Selur** [comment]
Any news (regarding ffmpeg integration or pipe support) ?

**Comment by zhangxianguo** [comment]
@Selur 
please email to zxgapple0001@163.com, we can give you the libs and ffmpeg wrappers, which you can use to produce a ffmpeg-libksc265. To protect our codec, the libs has expiration time.
